File: src/apps/daemon/internal/networking/proxyconfig/mesh_socks5_bridge.go
package proxyconfig

import (
	"bytes"
	"encoding/binary"
	"errors"
	"net"
	"sync"
)

// Socks5RoutingDecision represents routing disposition
type Socks5RoutingDecision int

const (
	Socks5RouteDirect Socks5RoutingDecision = iota
	Socks5RouteMeshExit
	Socks5RouteBlocked
)

// MeshExitTarget represents chosen mesh exit node
type MeshExitTarget struct {
	NodeID    string
	VirtualIP net.IP
	Active    bool
}

// MeshSocks5Bridge bridges local SOCKS5 clients to mesh exit nodes
type MeshSocks5Bridge struct {
	mu          sync.RWMutex
	ListenPort  uint16
	ExitTarget  *MeshExitTarget
	Credentials map[string]string
}

// NewMeshSocks5Bridge creates a bridge instance
func NewMeshSocks5Bridge(port uint16) *MeshSocks5Bridge {
	return &MeshSocks5Bridge{
		ListenPort:  port,
		Credentials: make(map[string]string),
	}
}

// SetExitTarget assigns the mesh exit target
func (b *MeshSocks5Bridge) SetExitTarget(target *MeshExitTarget) {
	b.mu.Lock()
	defer b.mu.Unlock()
	b.ExitTarget = target
}

// AddUser adds username and password authentication
func (b *MeshSocks5Bridge) AddUser(username, password string) {
	b.mu.Lock()
	defer b.mu.Unlock()
	b.Credentials[username] = password
}

// Authenticate verifies user credentials
func (b *MeshSocks5Bridge) Authenticate(user, pass string) bool {
	b.mu.RLock()
	defer b.mu.RUnlock()
	if len(b.Credentials) == 0 {
		return true
	}
	expected, ok := b.Credentials[user]
	return ok && expected == pass
}

// EvaluateRoute determines if request routes through mesh exit node
func (b *MeshSocks5Bridge) EvaluateRoute(host string, port uint16) (Socks5RoutingDecision, net.IP) {
	b.mu.RLock()
	defer b.mu.RUnlock()

	if port == 0 {
		return Socks5RouteBlocked, nil
	}

	if host == "localhost" || host == "127.0.0.1" {
		return Socks5RouteDirect, nil
	}

	if b.ExitTarget != nil && b.ExitTarget.Active {
		return Socks5RouteMeshExit, b.ExitTarget.VirtualIP
	}

	return Socks5RouteDirect, nil
}

// ParseGreeting checks SOCKS5 greeting methods
func (b *MeshSocks5Bridge) ParseGreeting(data []byte) (byte, error) {
	b.mu.RLock()
	defer b.mu.RUnlock()

	if len(data) < 2 {
		return 0xFF, errors.New("buffer too short")
	}
	if data[0] != 0x05 {
		return 0xFF, errors.New("invalid SOCKS version")
	}

	nmethods := int(data[1])
	if len(data) < 2+nmethods {
		return 0xFF, errors.New("incomplete methods list")
	}

	methods := data[2 : 2+nmethods]
	if len(b.Credentials) == 0 {
		if bytes.Contains(methods, []byte{0x00}) {
			return 0x00, nil // NO AUTH
		}
		return 0xFF, errors.New("client does not support NO_AUTH")
	}

	if bytes.Contains(methods, []byte{0x02}) {
		return 0x02, nil // USERNAME/PASSWORD
	}
	return 0xFF, errors.New("client does not support USER/PASSWORD AUTH")
}

// CraftReply generates SOCKS5 response packet
func (b *MeshSocks5Bridge) CraftReply(rep byte, bindIP net.IP, bindPort uint16) []byte {
	buf := new(bytes.Buffer)
	buf.WriteByte(0x05)
	buf.WriteByte(rep)
	buf.WriteByte(0x00) // Reserved

	v4 := bindIP.To4()
	if v4 != nil {
		buf.WriteByte(0x01) // IPv4
		buf.Write(v4)
	} else {
		buf.WriteByte(0x04) // IPv6
		buf.Write(bindIP.To16())
	}

	var portBytes [2]byte
	binary.BigEndian.PutUint16(portBytes[:], bindPort)
	buf.Write(portBytes[:])

	return buf.Bytes()
}

File: src/packages/lumicore/src/proxy/mesh_socks5_bridge.rs
// Pure Rust implementation: Mesh SOCKS5 Exit Node Bridge

use std::collections::HashMap;
use std::net::{Ipv4Addr, SocketAddr};

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Socks5AuthMethod {
    NoAuth,
    UsernamePassword(String, String),
}

#[derive(Debug, Clone)]
pub struct MeshExitNodeTarget {
    pub node_id: String,
    pub virtual_ip: Ipv4Addr,
    pub auth_token: Option<String>,
    pub active: bool,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum RouteDisposition {
    RouteViaMesh(Ipv4Addr),
    DirectBypass,
    Blocked(String),
}

pub struct MeshSocks5Bridge {
    pub listen_port: u16,
    pub target_exit_node: Option<MeshExitNodeTarget>,
    pub credentials: HashMap<String, String>,
    pub allowed_subnets: Vec<String>,
}

impl MeshSocks5Bridge {
    pub fn new(listen_port: u16) -> Self {
        Self {
            listen_port,
            target_exit_node: None,
            credentials: HashMap::new(),
            allowed_subnets: Vec::new(),
        }
    }

    pub fn set_exit_node(&mut self, target: MeshExitNodeTarget) {
        self.target_exit_node = Some(target);
    }

    pub fn add_user(&mut self, username: impl Into<String>, password: impl Into<String>) {
        self.credentials.insert(username.into(), password.into());
    }

    pub fn authenticate(&self, user: &str, pass: &str) -> bool {
        if self.credentials.is_empty() {
            return true;
        }
        self.credentials.get(user).map(|p| p == pass).unwrap_or(false)
    }

    pub fn evaluate_route(&self, target_host: &str, target_port: u16) -> RouteDisposition {
        if target_port == 0 {
            return RouteDisposition::Blocked("Invalid destination port".to_string());
        }

        // Loopback / RFC1918 bypass checks
        if target_host == "localhost" || target_host == "127.0.0.1" {
            return RouteDisposition::DirectBypass;
        }

        if let Some(ref exit) = self.target_exit_node {
            if exit.active {
                return RouteDisposition::RouteViaMesh(exit.virtual_ip);
            }
        }

        RouteDisposition::DirectBypass
    }

    pub fn parse_socks5_greeting(&self, buffer: &[u8]) -> Result<u8, &'static str> {
        if buffer.len() < 2 {
            return Err("Greeting buffer too short");
        }
        if buffer[0] != 0x05 {
            return Err("Invalid SOCKS version (expected 0x05)");
        }
        let nmethods = buffer[1] as usize;
        if buffer.len() < 2 + nmethods {
            return Err("Incomplete methods list");
        }

        let methods = &buffer[2..2 + nmethods];
        if self.credentials.is_empty() {
            if methods.contains(&0x00) {
                Ok(0x00) // NO AUTHENTICATION REQUIRED
            } else {
                Err("Client does not support NO_AUTH")
            }
        } else {
            if methods.contains(&0x02) {
                Ok(0x02) // USERNAME/PASSWORD
            } else {
                Err("Client does not support USER/PASS AUTH")
            }
        }
    }

    pub fn craft_greeting_response(&self, selected_method: u8) -> [u8; 2] {
        [0x05, selected_method]
    }

    pub fn craft_reply(
        &self,
        rep_code: u8,
        bound_addr: SocketAddr,
    ) -> Vec<u8> {
        let mut reply = vec![0x05, rep_code, 0x00];
        match bound_addr {
            SocketAddr::V4(v4) => {
                reply.push(0x01); // IPv4
                reply.extend_from_slice(&v4.ip().octets());
                reply.extend_from_slice(&v4.port().to_be_bytes());
            }
            SocketAddr::V6(v6) => {
                reply.push(0x04); // IPv6
                reply.extend_from_slice(&v6.ip().octets());
                reply.extend_from_slice(&v6.port().to_be_bytes());
            }
        }
        reply
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_socks5_greeting_and_auth() {
        let mut bridge = MeshSocks5Bridge::new(1080);
        
        // No auth configured -> accepts 0x00
        let greeting = [0x05, 0x01, 0x00];
        assert_eq!(bridge.parse_socks5_greeting(&greeting), Ok(0x00));
        assert_eq!(bridge.craft_greeting_response(0x00), [0x05, 0x00]);

        // With user configured -> requires 0x02
        bridge.add_user("alice", "secret123");
        assert_eq!(bridge.parse_socks5_greeting(&greeting), Err("Client does not support USER/PASS AUTH"));
        
        let greeting_with_auth = [0x05, 0x02, 0x00, 0x02];
        assert_eq!(bridge.parse_socks5_greeting(&greeting_with_auth), Ok(0x02));
        assert!(bridge.authenticate("alice", "secret123"));
        assert!(!bridge.authenticate("alice", "wrongpass"));
    }

    #[test]
    fn test_route_evaluation_and_reply_crafting() {
        let mut bridge = MeshSocks5Bridge::new(1080);
        let exit = MeshExitNodeTarget {
            node_id: "exit-de-1".to_string(),
            virtual_ip: Ipv4Addr::new(100, 64, 0, 99),
            auth_token: None,
            active: true,
        };
        bridge.set_exit_node(exit);

        // Localhost bypasses
        assert_eq!(bridge.evaluate_route("127.0.0.1", 80), RouteDisposition::DirectBypass);

        // External goes via mesh exit
        assert_eq!(
            bridge.evaluate_route("1.1.1.1", 443),
            RouteDisposition::RouteViaMesh(Ipv4Addr::new(100, 64, 0, 99))
        );

        // Craft reply packet
        let addr: SocketAddr = "100.64.0.1:1080".parse().unwrap();
        let reply = bridge.craft_reply(0x00, addr);
        assert_eq!(reply[0], 0x05); // SOCKS5
        assert_eq!(reply[1], 0x00); // SUCCESS
        assert_eq!(reply[3], 0x01); // IPv4
    }
}
